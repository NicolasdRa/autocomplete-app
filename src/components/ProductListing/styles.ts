import { makeStyles, Theme, createStyles } from '@material-ui/core/styles';

export const useStyles = makeStyles((theme: Theme) =>
	createStyles({
		container: {
			width: '90vw',
			margin: '0 auto',
			overflowX: 'auto',
		},

		toolbar: {
			display: 'flex',
			flexDirection: 'row',
			justifyContent: 'space-between',
			alignItems: 'center',
			marginTop: '.5rem',

			[theme.breakpoints.down('sm')]: {
				margin: 0,
			},
		},

		title: {
			marginLeft: '.8rem',
			padding: '2rem 0 1rem 1rem',

			[theme.breakpoints.down('sm')]: {
				padding: '1rem 0 0 0',
			},
		},

		filters: {
			display: 'flex',
			flexDirection: 'row',
			justifyContent: 'flex-end',
			alignItems: 'center',
			padding: '2rem 0 1rem 0',

			[theme.breakpoints.down('sm')]: {
				flexWrap: 'wrap',
				justifyContent: 'start',
				padding: '1rem 0 0 0',
			},
		},

		select: {
			width: '9rem',
			margin: '0 1rem',

			[theme.breakpoints.down('sm')]: {
				width: '50%',
				marginLeft: 0,
				marginBottom: '1rem',
			},
		},

		checkbox: {
			margin: '0 .5rem',

			[theme.breakpoints.down('sm')]: {
				width: '30%',
				marginLeft: '2rem',
				marginBottom: '1rem',
			},
		},
	}),
);
