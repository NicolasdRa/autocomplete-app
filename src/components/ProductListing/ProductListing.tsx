import { useState, useEffect } from 'react';

import {
	Box,
	FormControl,
	FormGroup,
	FormControlLabel,
	Checkbox,
	InputLabel,
	Select,
	MenuItem,
	Typography,
} from '@material-ui/core';

import { readString } from 'react-papaparse';
import csvFile from '../../data/products.csv';

import { Search } from '../Search/Search';
import { CustomTable } from '../CustomTable/CustomTable';

import { useStyles } from './styles';

export const ProductListing: React.FC = () => {
	const classes = useStyles();

	const [dataArray, setDataArray] = useState<any[]>([]);
	const [filteredArray, setFilteredArray] = useState<any[]>([]);
	const [queryString, setQueryString] = useState('');
	const [gender, setGender] = useState('');
	const [checked, setChecked] = useState(false);

	useEffect(() => {
		readString(csvFile, {
			delimiter: '',
			newline: '',
			header: true,
			skipEmptyLines: true,
			complete: (data) => {
				console.log('Parsing complete', data);
				setDataArray(data.data);
			},
			download: true,
			error: (error, file) => {
				console.log('Error while parsing:', error, file);
			},
		});
	}, []);

	useEffect(() => {
		dataArray.length > 0 && setFilteredArray(dataArray);
	}, [dataArray]);

	// search filter
	const search = (data: any[]): any[] => {
		if (data.length > 0 && queryString !== '') {
			return data.filter((item) =>
				item.title.toLowerCase().includes(queryString),
			);
		}
		return data;
	};

	// on Sale filter
	const filterByOnSale = (data: any[]): any[] => {
		if (checked) {
			return data.filter((item) => item.sale_price < item.price);
		}
		return data;
	};

	// gender filter
	const filterByGender = (data: any[]): any[] => {
		if (gender !== 'unisex') {
			return data.filter((item) => item.gender === gender);
		}
		return data;
	};

	useEffect(() => {
		const searchFiltered = search(dataArray);
		const saleFiltered = filterByOnSale(searchFiltered);
		const genderFiltered = filterByGender(saleFiltered);

		setFilteredArray(genderFiltered);
	}, [dataArray, queryString, gender, checked]);

	const handleChangeQuery = (e: any) => {
		setQueryString(e.target.value.toLowerCase());
	};

	const handleChangeSaleFilter = (e: any) => {
		setChecked(e.target.checked);
	};

	const handleChangeGenderFilter = (e: any) => {
		setGender(e.target.value);
	};

	return (
		<Box>
			<Box className={classes.toolbar}>
				<Typography variant='h5' className={classes.title}>
					Products
				</Typography>
				<Box className={classes.inputs}>
					<Search handleChangeQuery={handleChangeQuery} />
					<FormGroup row className={classes.checkbox}>
						<FormControlLabel
							control={
								<Checkbox
									checked={checked}
									onChange={handleChangeSaleFilter}
									color='primary'
								/>
							}
							label='On Sale'
						/>
					</FormGroup>
					<FormControl
						variant='outlined'
						className={classes.formControl}>
						<InputLabel id='demo-simple-select-outlined-label'>
							gender
						</InputLabel>
						<Select
							labelId='demo-simple-select-outlined-label'
							id='demo-simple-select-outlined'
							value={gender}
							onChange={handleChangeGenderFilter}
							label='Gender'>
							<MenuItem value=''>
								<em>None</em>
							</MenuItem>
							<MenuItem value={'female'}>female</MenuItem>
							<MenuItem value={'male'}>male</MenuItem>
							<MenuItem value={'unisex'}>unisex</MenuItem>
						</Select>
					</FormControl>
				</Box>
			</Box>

			{dataArray.length > 0 && (
				<CustomTable data={filteredArray.slice(0, 99)} />
			)}
		</Box>
	);
};
