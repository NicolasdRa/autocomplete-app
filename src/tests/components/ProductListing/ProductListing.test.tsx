import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { ProductListing } from '../../../components/ProductListing/ProductListing';

const mockData = [
	{
		title: 'test object',
		gtin: '1',
		gender: 'unisex',
		price: '1 euro',
		sale_price: '0.5 euro',
		image_link: 'testImg.jpg',
		additional_image_link: 'testImg1.jpg, testImg2.jpg',
	},

	{
		title: 'test object2',
		gtin: '1',
		gender: 'female',
		price: '2 euro',
		sale_price: '0.5 euro',
		image_link: 'testImg.jpg',
		additional_image_link: 'testImg1.jpg, testImg2.jpg',
	},

	{
		title: 'test object3',
		gtin: '1',
		gender: 'male',
		price: '3 euro',
		sale_price: '0.5 euro',
		image_link: 'testImg.jpg',
		additional_image_link: 'testImg1.jpg, testImg2.jpg',
	},
];

describe('tests on <ProductListing/>', () => {
	const wrapper = shallow(<ProductListing data={mockData} />);

	test('should render <ProductListing/> correctly', () => {
		expect(wrapper).toMatchSnapshot();
	});
});
