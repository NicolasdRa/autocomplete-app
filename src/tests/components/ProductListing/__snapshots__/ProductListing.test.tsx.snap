// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on <ProductListing/> should render <ProductListing/> correctly 1`] = `
<Styled(MuiBox)
  className="makeStyles-container-1"
>
  <Styled(MuiBox)
    className="makeStyles-toolbar-2"
  >
    <Hidden
      smDown={true}
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-title-3"
        variant="h5"
      >
        Products
      </WithStyles(ForwardRef(Typography))>
    </Hidden>
    <Styled(MuiBox)
      className="makeStyles-filters-4"
    >
      <Search
        handleChangeQuery={[Function]}
      />
      <WithStyles(ForwardRef(FormControl))
        className="makeStyles-select-5"
        variant="outlined"
      >
        <WithStyles(ForwardRef(InputLabel))
          id="demo-simple-select-outlined-label"
        >
          Filter by gender
        </WithStyles(ForwardRef(InputLabel))>
        <WithStyles(ForwardRef(Select))
          id="demo-simple-select-outlined"
          labelId="demo-simple-select-outlined-label"
          labelWidth={100}
          onChange={[Function]}
          value=""
        >
          <WithStyles(ForwardRef(MenuItem))
            value=""
          >
            <em>
              None
            </em>
          </WithStyles(ForwardRef(MenuItem))>
          <WithStyles(ForwardRef(MenuItem))
            value="female"
          >
            female
          </WithStyles(ForwardRef(MenuItem))>
          <WithStyles(ForwardRef(MenuItem))
            value="male"
          >
            male
          </WithStyles(ForwardRef(MenuItem))>
          <WithStyles(ForwardRef(MenuItem))
            value="unisex"
          >
            unisex
          </WithStyles(ForwardRef(MenuItem))>
        </WithStyles(ForwardRef(Select))>
      </WithStyles(ForwardRef(FormControl))>
      <WithStyles(ForwardRef(FormGroup))
        className="makeStyles-checkbox-6"
        row={true}
      >
        <WithStyles(ForwardRef(FormControlLabel))
          control={
            <WithStyles(ForwardRef(Checkbox))
              checked={false}
              color="primary"
              onChange={[Function]}
            />
          }
          label="On Sale"
        />
      </WithStyles(ForwardRef(FormGroup))>
      <Hidden
        smUp={true}
      >
        <WithStyles(ForwardRef(Typography))
          className="makeStyles-title-3"
          variant="h5"
        >
          Products
        </WithStyles(ForwardRef(Typography))>
      </Hidden>
    </Styled(MuiBox)>
  </Styled(MuiBox)>
  <CustomTable
    data={
      Array [
        Object {
          "additional_image_link": "testImg1.jpg, testImg2.jpg",
          "gender": "unisex",
          "gtin": "1",
          "image_link": "testImg.jpg",
          "price": "1 euro",
          "sale_price": "0.5 euro",
          "title": "test object",
        },
        Object {
          "additional_image_link": "testImg1.jpg, testImg2.jpg",
          "gender": "female",
          "gtin": "1",
          "image_link": "testImg.jpg",
          "price": "2 euro",
          "sale_price": "0.5 euro",
          "title": "test object2",
        },
        Object {
          "additional_image_link": "testImg1.jpg, testImg2.jpg",
          "gender": "male",
          "gtin": "1",
          "image_link": "testImg.jpg",
          "price": "3 euro",
          "sale_price": "0.5 euro",
          "title": "test object3",
        },
      ]
    }
  />
</Styled(MuiBox)>
`;
